Libplanet.Net.Tests/Consensus/ByzantineTest.cs

```diff
@@ -0,0 +1,160 @@
+            const string outputTemplate =
+                "{Timestamp:HH:mm:ss:ffffffZ} - {Message} {Exception}";
+            Log.Logger = new LoggerConfiguration()
+                .MinimumLevel.Debug()
+                .WriteTo.TestOutput(output, outputTemplate: outputTemplate)
+                .CreateLogger()
+                .ForContext<ByzantineTest>();
+
+            _logger = Log.ForContext<ByzantineTest>();
+        }
+
+        [Fact(Timeout = Timeout)]
+        public async Task DoubleVote()
+        {
+            // If a byzantine node sends two different votes in a round, by the MessageLog
+            // specification, one of the vote is ignored and keep the total quorum remains the same.
+
+            // Due to giving a BroadcastMessage delegation, here Context<T> is created in
+            // first and defined null. The initialization will be done in later.
+            const int validatorCount = 4;
+            Gossip[]? gossips = null;
+
+            var peerList = Enumerable.Range(0, 4)
+                .Select(x =>
+                    new BoundPeer(
+                        TestUtils.PrivateKeys[x].PublicKey,
+                        new DnsEndPoint("127.0.0.1", 6000 + x)))
+                .ToArray();
+
+            (BlockChain<DumbAction> BlockChain, Context<DumbAction> Context)[]?
+                validators = null;
+            var newHeightStarted = new AsyncAutoResetEvent[validatorCount];
+            for (int i = 0; i < validatorCount; ++i)
+            {
+                newHeightStarted[i] = new AsyncAutoResetEvent();
+            }
+
+            gossips = Enumerable.Range(0, 4)
+                .Select(x => TestUtils.CreateGossip(
+                    content =>
+                    {
+                        if (content is ConsensusMsg consensusMsg)
+                        {
+                            // ReSharper disable once AccessToModifiedClosure
+                            validators![x].Context.ProduceMessage(consensusMsg);
+                        }
+                    },
+                    TestUtils.PrivateKeys[x],
+                    6000 + x,
+                    peerList))
+                .ToArray();
+
+            validators = Enumerable.Range(0, validatorCount)
+                .Select(x =>
+                    TestUtils.CreateDummyContext(
+                        1,
+                        privateKey: TestUtils.PrivateKeys[x],
+                        broadcastMessage: gossips[x].AddMessage))
+                .ToArray();
+
+            try
+            {
+                // Setup next height listener.
+                for (int i = 0; i < validatorCount; i++)
+                {
+                    var nodeNo = i;
+                    validators[i].Context.StateChanged += (sender, mrs) =>
+                    {
+                        if (mrs.Step == Step.EndCommit)
+                        {
+                            newHeightStarted[nodeNo].Set();
+                        }
```

Comment:
State change to `EndCommit` doesn't assure `newHeight` has been started, due to the existence of `newHeightDelay`.